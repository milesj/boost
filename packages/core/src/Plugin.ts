/**
 * @copyright   2017-2018, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import Module from './Module';
import CoreTool from './Tool';

export const DEFAULT_PLUGIN_PRIORITY: number = 100;

export default class Plugin<Tool extends CoreTool, Options = {}> extends Module<Options> {
  priority: number = DEFAULT_PLUGIN_PRIORITY;

  // @ts-ignore Set after instantiation
  tool: Tool;

  bootstrap() {}
}
