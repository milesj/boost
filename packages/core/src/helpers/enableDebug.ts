/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import debug from 'debug';

export default function enableDebug(namespace: string) {
  const { DEBUG } = process.env;
  const flag = `${namespace}:*`;

  if (DEBUG) {
    if (DEBUG.includes(flag)) {
      return;
    }

    process.env.DEBUG += `,${flag}`;
  } else {
    process.env.DEBUG = flag;
  }

  debug.enabled(flag);
}
