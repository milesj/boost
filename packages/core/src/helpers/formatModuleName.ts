/**
 * @copyright   2017-2018, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

/**
 * Combine app and plugin names into a Node/NPM applicable module name.
 */
export default function formatModuleName(
  appName: string,
  pluginType: string,
  name: string,
  scoped: boolean = false,
): string {
  const moduleName = name.toLowerCase().replace(`${pluginType}:`, '');

  if (scoped) {
    return `@${appName}/${pluginType}-${moduleName}`;
  }

  return `${appName}-${pluginType}-${moduleName}`;
}
