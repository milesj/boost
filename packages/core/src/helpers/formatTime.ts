import prettyMs from 'pretty-ms';

export default function formatTime(ms: number): string {
  if (!Number.isFinite(ms) || ms === 0) {
    return '0s';
  }

  return prettyMs(ms, { keepDecimalsOnWholeSeconds: true });
}
