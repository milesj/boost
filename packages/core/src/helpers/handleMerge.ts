/**
 * @copyright   2017-2018, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

/**
 * Handle special cases when merging 2 configuration values.
 * If the target and source are both arrays, concatenate them.
 */
export default function handleMerge(target: any, source: any): any {
  if (Array.isArray(target) && Array.isArray(source)) {
    return Array.from(new Set([...target, ...source]));
  }

  // Defer to lodash
  return undefined;
}
