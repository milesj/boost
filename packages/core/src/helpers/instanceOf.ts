/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { Constructor } from '../types';

/**
 * Native `instanceof` checks are problematic, as cross realm checks fail.
 * They will also fail when comparing against source and built files.
 * So emulate an `instanceof` check by comparing constructor names.
 */
export default function instanceOf<T = any>(object: any, contract: Constructor<T>): object is T {
  if (!object || typeof object !== 'object') {
    return false;
  }

  if (object instanceof contract) {
    return true;
  }

  let current = object;

  while (current) {
    if (current.constructor.name === contract.name) {
      return true;
    }

    current = Object.getPrototypeOf(current);
  }

  return false;
}
