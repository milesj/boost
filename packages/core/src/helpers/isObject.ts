/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

/**
 * Return true if the value is a literal object.
 */
export default function isObject(value: any): boolean {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}
