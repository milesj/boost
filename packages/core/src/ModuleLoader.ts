import {
  instanceOf,
  isObject,
  requireModule,
  Constructor,
  ConcreteConstructor,
  Path,
} from '@boost/common';
import { createDebugger, Debugger } from '@boost/debug';
import { color } from '@boost/internal';
import formatModuleName from './helpers/formatModuleName';
import Tool from './Tool';

export interface OptionsObject {
  [key: string]: any;
}

export default class ModuleLoader<Tm> {
  contract: Constructor<Tm> | null = null;

  debug: Debugger;

  scopes: string[];

  tool: Tool<any>;

  typeName: string;

  constructor(
    tool: Tool<any>,
    typeName: string,
    contract: Constructor<Tm> | null = null,
    scopes: string[] = [],
  ) {
    this.contract = contract;
    this.debug = createDebugger(`${typeName}-loader`);
    this.scopes = scopes;
    this.tool = tool;
    this.typeName = typeName;
  }

  /**
   * Import a class definition from a Node module and instantiate the class
   * with the provided options object.
   */
  importModule(name: string, args: any[] = []): Tm {
    const { typeName } = this;
    const { appName, scoped } = this.tool.options;

    // Determine modules to attempt to load
    const modulesToAttempt = [];
    let isFilePath = false;
    let importedModule: any = null;
    let moduleName;

    // File path
    if (name.match(/^\.|\/|\\|[A-Z]:/u)) {
      this.debug('Locating %s from path %s', typeName, color.filePath(name));

      modulesToAttempt.push(Path.create(name).path());
      isFilePath = true;

      // Module name
    } else {
      this.debug('Locating %s module %s', typeName, color.moduleName(name));

      if (scoped) {
        modulesToAttempt.push(formatModuleName(appName, typeName, name, true));
      }

      modulesToAttempt.push(formatModuleName(appName, typeName, name));

      // Additional scopes to load
      this.scopes.forEach(otherScope => {
        modulesToAttempt.push(
          formatModuleName(otherScope, typeName, name, true),
          formatModuleName(otherScope, typeName, name),
        );
      });

      this.debug('Resolving in order: %s', modulesToAttempt.join(', '));
    }

    modulesToAttempt.some(modName => {
      try {
        importedModule = requireModule(modName);
        moduleName = modName;

        return true;
      } catch (error) {
        if (error.message.startsWith(`Cannot find module '${modName}'`)) {
          this.debug('Failed to import module: %s', error.message);

          return false;
        }

        // Unknown error occurred, abort process
        throw error;
      }
    });

    if (!importedModule || !moduleName) {
      throw new Error(
        this.tool.msg('errors:moduleImportFailed', {
          modules: modulesToAttempt.join(', '),
          typeName,
        }),
      );
    }

    if (!this.contract) {
      return importedModule;
    }

    // An instance was returned instead of the class definition
    if (instanceOf(importedModule, this.contract)) {
      throw new TypeError(
        this.tool.msg('errors:moduleClassInstanceExported', {
          appName,
          moduleName,
          typeName,
        }),
      );
    } else if (typeof importedModule !== 'function') {
      throw new TypeError(this.tool.msg('errors:moduleClassDefRequired', { moduleName, typeName }));
    }

    const ModuleClass = importedModule as ConcreteConstructor<Tm>;
    const module = new ModuleClass(...args);

    if (!instanceOf(module, this.contract)) {
      throw new TypeError(
        this.tool.msg('errors:moduleExportInvalid', {
          moduleName,
          typeName,
        }),
      );
    }

    if (isFilePath) {
      this.debug('Found with file path %s', color.filePath(moduleName));
    } else {
      this.debug('Found with module %s', color.moduleName(moduleName));

      (module as any).name = name;
      (module as any).moduleName = moduleName;
    }

    return module;
  }

  /**
   * If loading from an object, extract the module name and use the remaining object
   * as options for the class instance.
   */
  importModuleFromOptions(baseOptions: OptionsObject, args: any[] = []): Tm {
    const { typeName } = this;
    const options = { ...baseOptions };
    const module = options[typeName];

    delete options[typeName];

    if (!module || typeof module !== 'string') {
      throw new TypeError(this.tool.msg('errors:moduleOptionMissingKey', { typeName }));
    }

    const nextArgs = [...args];

    if (nextArgs.length === 0) {
      nextArgs.push(options);
    } else if (isObject<object>(nextArgs[0])) {
      nextArgs[0] = {
        ...nextArgs[0],
        ...options,
      };
    }

    return this.importModule(module, nextArgs);
  }

  /**
   * Load and or instantiate a module for the `typeName` configuration property.
   * If a class instance, use directly. If a string, attempt to load and
   * instantiate from a module. If an object, extract the name and run the previous.
   */
  loadModule(module: string | OptionsObject | Tm, args: any[] = []): Tm {
    if (this.contract && instanceOf(module, this.contract)) {
      return module as Tm;
    } else if (typeof module === 'string') {
      return this.importModule(module, args);
    } else if (isObject(module)) {
      return this.importModuleFromOptions(module, args);
    }

    throw new TypeError(this.tool.msg('errors:moduleTypeInvalid', { typeName: this.typeName }));
  }

  /**
   * Load multiple modules.
   */
  loadModules(modules: (string | OptionsObject | Tm)[] = [], args: any[] = []): Tm[] {
    return modules.map(module => this.loadModule(module, args));
  }
}
