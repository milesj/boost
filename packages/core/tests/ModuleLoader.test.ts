import Plugin from '../src/Plugin';
import ModuleLoader from '../src/ModuleLoader';
import { getFixturePath, getTestRoot, copyFixtureToMock, createTestTool } from './helpers';

function createPlugin(name: string, options: any = {}): Plugin {
  const plugin = new Plugin(options);
  plugin.name = name;
  plugin.moduleName = `test-boost-plugin-${name}`;

  return plugin;
}

describe('ModuleLoader', () => {
  let loader: ModuleLoader<any>;
  let fixtures: (() => void)[] = [];

  beforeEach(() => {
    loader = new ModuleLoader(
      createTestTool({
        root: getTestRoot(),
      }),
      'plugin',
      Plugin,
    );

    fixtures = [];
  });

  afterEach(() => {
    fixtures.forEach(remove => remove());
  });

  describe('importModule()', () => {
    it('errors for missing node module', () => {
      expect(() => {
        loader.importModule('missing');
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors for missing node module (with scope)', () => {
      expect(() => {
        loader.tool.options.scoped = true;
        loader.importModule('missing');
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors for missing node module (with boost)', () => {
      expect(() => {
        loader.loadBoostModules = true;
        loader.tool.options.scoped = true;
        loader.importModule('missing');
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if a non-function is exported', () => {
      fixtures.push(copyFixtureToMock('plugin-exported-nonfunc', 'test-boost-plugin-nonfunc'));

      expect(() => {
        loader.importModule('nonfunc');
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if a non-plugin is exported', () => {
      fixtures.push(copyFixtureToMock('plugin-exported-func', 'test-boost-plugin-func'));

      expect(() => {
        loader.importModule('func');
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if a plugin instance is exported', () => {
      fixtures.push(copyFixtureToMock('plugin-exported-instance', 'test-boost-plugin-instance'));

      expect(() => {
        loader.importModule('instance');
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if a non-plugin instance is exported', () => {
      fixtures.push(
        copyFixtureToMock('plugin-exported-nonplugin-instance', 'test-boost-plugin-nonplugin'),
      );

      expect(() => {
        loader.importModule('nonplugin');
      }).toThrowErrorMatchingSnapshot();
    });

    it('imports and instantiates a plugin', () => {
      fixtures.push(
        copyFixtureToMock('plugin-exported-definition', 'test-boost-plugin-definition'),
      );

      const plugin = loader.importModule('definition');

      expect(plugin).toBeInstanceOf(Plugin);
      expect(plugin.name).toBe('definition');
      expect(plugin.moduleName).toBe('test-boost-plugin-definition');
    });

    it('imports and returns a theme object', () => {
      fixtures.push(copyFixtureToMock('theme-exported-object', 'test-boost-theme-object'));

      loader = new ModuleLoader(
        createTestTool({
          root: getTestRoot(),
        }),
        'theme',
      );

      const palette = loader.importModule('object');

      expect(palette).toEqual({
        default: 'white',
        failure: 'red',
        pending: 'gray',
        success: 'green',
        warning: 'yellow',
      });
    });

    it('imports using a file path', () => {
      fixtures.push(
        copyFixtureToMock('plugin-exported-definition', 'test-boost-plugin-definition'),
      );

      const plugin = loader.importModule(getFixturePath('plugin-exported-definition'));

      expect(plugin).toBeInstanceOf(Plugin);
      expect(plugin.name).toBe('');
      expect(plugin.moduleName).toBe('');
    });

    it('can customize type name', () => {
      fixtures.push(copyFixtureToMock('plugin-exported-definition', 'test-boost-addon-definition'));

      loader.typeName = 'addon';

      const plugin = loader.importModule('definition');

      expect(plugin.name).toBe('definition');
      expect(plugin.moduleName).toBe('test-boost-addon-definition');
    });

    it('supports scopeds', () => {
      fixtures.push(copyFixtureToMock('plugin-exported-definition', '@test-boost/plugin-scoped'));
      loader.tool.options.scoped = true;

      const plugin = loader.importModule('scoped');

      expect(plugin.name).toBe('scoped');
      expect(plugin.moduleName).toBe('@test-boost/plugin-scoped');
    });
  });

  describe('importModuleFromOptions()', () => {
    it('errors if key doesnt exist', () => {
      expect(() => {
        loader.importModuleFromOptions({});
      }).toThrowErrorMatchingSnapshot();
    });

    it('errors if key is not a string', () => {
      expect(() => {
        loader.importModuleFromOptions({ plugin: [] });
      }).toThrowErrorMatchingSnapshot();
    });

    it('imports and instantiates a plugin with options', () => {
      fixtures.push(
        copyFixtureToMock('plugin-exported-definition', 'test-boost-plugin-definition-with-opts'),
      );

      const plugin = loader.importModuleFromOptions({
        foo: 'bar',
        plugin: 'definition-with-opts',
      });

      expect(plugin.options).toEqual({ foo: 'bar' });
      expect(plugin.name).toBe('definition-with-opts');
      expect(plugin.moduleName).toBe('test-boost-plugin-definition-with-opts');
    });

    it('merges options correctly', () => {
      fixtures.push(
        copyFixtureToMock('plugin-exported-definition', 'test-boost-plugin-definition-merge-opts'),
      );

      const plugin = loader.importModuleFromOptions(
        {
          foo: 'bar',
          plugin: 'definition-with-opts',
        },
        [{ foo: 'wtf', baz: 123 }],
      );

      expect(plugin.options).toEqual({ foo: 'bar', baz: 123 });
    });
  });

  describe('loadModules()', () => {
    it('errors if an unsupported value is passed', () => {
      expect(() => {
        loader.loadModules([123]);
      }).toThrowErrorMatchingSnapshot();
    });

    it('returns an empty array if no values', () => {
      expect(loader.loadModules()).toEqual([]);
      expect(loader.loadModules([])).toEqual([]);
    });

    it('supports class instances', () => {
      const plugins = [createPlugin('foo'), createPlugin('bar'), createPlugin('baz')];

      expect(loader.loadModules(plugins)).toEqual(plugins);
    });

    it('supports string names', () => {
      fixtures.push(copyFixtureToMock('plugin-exported-definition', 'test-boost-plugin-foo'));
      fixtures.push(copyFixtureToMock('plugin-exported-definition', 'test-boost-plugin-bar'));
      fixtures.push(copyFixtureToMock('plugin-exported-definition', 'test-boost-plugin-baz'));

      expect(loader.loadModules(['foo', 'bar', 'baz'])).toEqual([
        createPlugin('foo'),
        createPlugin('bar'),
        createPlugin('baz'),
      ]);
    });

    it('supports options objects', () => {
      fixtures.push(copyFixtureToMock('plugin-exported-definition', 'test-boost-plugin-foo'));
      fixtures.push(copyFixtureToMock('plugin-exported-definition', 'test-boost-plugin-bar'));
      fixtures.push(copyFixtureToMock('plugin-exported-definition', 'test-boost-plugin-baz'));

      expect(loader.loadModules([{ plugin: 'foo' }, { plugin: 'bar' }, { plugin: 'baz' }])).toEqual(
        [createPlugin('foo'), createPlugin('bar'), createPlugin('baz')],
      );
    });

    it('supports all 3 at once', () => {
      fixtures.push(copyFixtureToMock('plugin-exported-definition', 'test-boost-plugin-bar'));
      fixtures.push(copyFixtureToMock('plugin-exported-definition', 'test-boost-plugin-baz'));

      expect(loader.loadModules([createPlugin('foo'), 'bar', { plugin: 'baz' }])).toEqual([
        createPlugin('foo'),
        createPlugin('bar'),
        createPlugin('baz'),
      ]);
    });

    it('pass options to string names', () => {
      fixtures.push(copyFixtureToMock('plugin-exported-definition', 'test-boost-plugin-foo'));

      expect(loader.loadModules(['foo'], [{ test: true }])).toEqual([
        createPlugin('foo', { test: true }),
      ]);
    });

    it('pass options to option objects', () => {
      fixtures.push(copyFixtureToMock('plugin-exported-definition', 'test-boost-plugin-foo'));

      expect(loader.loadModules([{ plugin: 'foo' }], [{ test: true }])).toEqual([
        createPlugin('foo', { test: true }),
      ]);
    });
  });
});
