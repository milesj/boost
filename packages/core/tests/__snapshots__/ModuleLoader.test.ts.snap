// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModuleLoader importModule() errors for missing node module (with boost) 1`] = `"Missing plugin. Attempted import in order: @test-boost/plugin-missing, test-boost-plugin-missing, @boost/plugin-missing, boost-plugin-missing"`;

exports[`ModuleLoader importModule() errors for missing node module (with scope) 1`] = `"Missing plugin. Attempted import in order: @test-boost/plugin-missing, test-boost-plugin-missing"`;

exports[`ModuleLoader importModule() errors for missing node module 1`] = `"Missing plugin. Attempted import in order: test-boost-plugin-missing"`;

exports[`ModuleLoader importModule() errors if a non-function is exported 1`] = `"Invalid plugin class definition exported from \\"test-boost-plugin-nonfunc\\"."`;

exports[`ModuleLoader importModule() errors if a non-plugin instance is exported 1`] = `"Invalid plugin class definition exported from \\"test-boost-plugin-nonplugin\\"."`;

exports[`ModuleLoader importModule() errors if a non-plugin is exported 1`] = `"Plugin exported from \\"test-boost-plugin-func\\" is invalid."`;

exports[`ModuleLoader importModule() errors if a plugin instance is exported 1`] = `"A plugin class instance was exported from \\"test-boost-plugin-instance\\". Test-boost requires a plugin class definition to be exported."`;

exports[`ModuleLoader importModuleFromOptions() errors if key doesnt exist 1`] = `"A \\"plugin\\" property must exist when loading through an options object."`;

exports[`ModuleLoader importModuleFromOptions() errors if key is not a string 1`] = `"A \\"plugin\\" property must exist when loading through an options object."`;

exports[`ModuleLoader loadModules() errors if an unsupported value is passed 1`] = `"Invalid plugin. Must be a class instance or a module that exports a class definition."`;
